### PR TITLE
TINY-6907: Added new mceListUpdate command

### PR DESCRIPTION
**Pull request by TheSpyder**
Related Ticket: TINY-6907

Description of Changes:

The list properties dialog now uses a new `SetListAttributes` command to set the list start attribute. I've also ensured the `mceListProps` command is registered in RTC mode and made use of it in the menu item instead of opening the dialog directly.

I didn't add a changelog for this because the list properties dialog doesn't seem to be documented yet; it's also an internal change that should have no impact on the feature.

Pre-checks:
* [x] ~~Changelog entry added~~
* [x] Tests have been added (if applicable)
* [x] ~~Branch prefixed with `feature/` for new features (if applicable)~~
* [x] License headers added on new files (if applicable)

Review:
* [x] Milestone set
* [x] Review comments resolved

GitHub issues (if applicable):

### CODE REVIEW

File: modules/tinymce/src/plugins/lists/main/ts/actions/Attributes.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+ * Copyright (c) Tiny Technologies, Inc. All rights reserved.
+ * Licensed under the LGPL or a commercial license.
+ * For LGPL see License.txt in the project root for license information.
+ * For commercial licenses see https://www.tiny.cloud/
+ */
+
+import { Obj } from '@ephox/katamari';
+import { getParentList } from '../core/Selection';
+
+export const setParentListAttributes = (editor, attrs) => {
```
Comment (lnewson) [inline modules/tinymce/src/plugins/lists/main/ts/actions/Attributes.ts:11]: Types?

File: modules/tinymce/src/plugins/lists/main/ts/api/Commands.ts [hunk 2]
```diff
@@ -5,17 +5,25 @@
...
 
 const queryListCommandState = (editor, listName) => {
   return () => {
-    const parentList = editor.dom.getParent(editor.selection.getStart(), 'UL,OL,DL');
+    const parentList = getParentList(editor);
     return parentList && parentList.nodeName === listName;
   };
 };
 
+const registerDialog = (editor) => {
```
Comment (lnewson) [inline modules/tinymce/src/plugins/lists/main/ts/api/Commands.ts:21]: Types?

File: modules/tinymce/src/plugins/lists/main/ts/api/Commands.ts [hunk 3]
```diff
@@ -43,8 +51,10 @@ const register = (editor) => {
     flattenListSelection(editor);
   });
 
-  editor.addCommand('mceListProps', () => {
-    Dialog.open(editor);
+  registerDialog(editor);
+
+  editor.addCommand('SetListAttributes', (ui, detail) => {
```
Comment (lnewson) [inline modules/tinymce/src/plugins/lists/main/ts/api/Commands.ts:56]: This should probably be called `mceSetListAttributes` to keep with the current naming convention (ie custom commands are prefixed with `mce`).

- [ ] Change command name to `mceSetListAttributes`

Comment (lnewson) [reply modules/tinymce/src/plugins/lists/main/ts/api/Commands.ts:56]: I'm also wondering if this maybe should be called `mceListMutation` (or similar) to go with the event of a similar name? If we do that, we probably need to change the data type to something like `{ attrs: Record<string, string> }`. That would then make this much more flexible for future potential, if say we needed to set styles or classes, etc...

File: modules/tinymce/src/plugins/lists/test/ts/browser/ListPropertiesTest.ts [hunk 4]
```diff
@@ -132,4 +132,52 @@ describe('browser.tinymce.plugins.lists.ListPropertiesTest', () => {
     await TinyUiActions.pWaitForUi(editor, '.tox-collection__item:contains("List properties"):not(.tox-collection__item--state-disabled)');
     TinyUiActions.clickOnMenu(editor, '.tox-mbtn:contains("Custom")');
   });
+
+  it('TINY-6907: List properties command is used to update the DOM', async () => {
+    const editor = hook.editor();
+
+    const blockCommand = (event) => {
```
Comment (lnewson) [inline modules/tinymce/src/plugins/lists/test/ts/browser/ListPropertiesTest.ts:139]: Types for event?

**Review by lnewson — APPROVED** [review]

**Comment by TheSpyder** [comment]
Sigh, I forgot our TypeScript settings don't warn me when I forget to add types. Sorry about that.

File: modules/tinymce/src/plugins/lists/main/ts/api/Commands.ts (hunk 3 above)
Comment (TheSpyder) [reply modules/tinymce/src/plugins/lists/main/ts/api/Commands.ts:56]: I have no attachment to the command name, I'm happy to go with whatever the team wants. I am also concerned about the scalability of this to later set CSS attributes; `ApplyUnorderedListStyle` and `ApplyOrderedListStyle` from `advlist` could be replaced with this command if we do that.

A sub-object seemed like overkill but is probably the best approach.

**Comment by lnewson** [comment]
> Sigh, I forgot our TypeScript settings don't warn me when I forget to add types. Sorry about that.

Yeah, I'd love to solve that... we were at ~10k errors in strict mode back in mid-2019 and we're now down to ~7k errors. That's still a long way to go before we can enable it though :cry:

**Comment by lnewson** [comment]
Oops, just remembered we'd only need `noImplicitAny` for that purpose... we have ~3.8k errors still to resolve by just enabling that check.

**Comment by TheSpyder** [comment]
It also didn't help that some of the code I was editing is missing types 😂 I'll add a few more in the files I edited.

File: modules/tinymce/src/plugins/lists/main/ts/api/Commands.ts (hunk 3 above)
Comment (TheSpyder) [reply modules/tinymce/src/plugins/lists/main/ts/api/Commands.ts:57]: We don't seem to use `Mutate` in our event names anywhere, how about `mceUpdateList`?

Comment (lnewson) [reply modules/tinymce/src/plugins/lists/main/ts/api/Commands.ts:57]: The event name is `ListMutation`, but it's used when changing the types mostly. That suggestion seems fine to me, though we do normally try to keep the list name first, eg `mceListUpdate`.

Comment (TheSpyder) [reply modules/tinymce/src/plugins/lists/main/ts/api/Commands.ts:57]: ah I was only looking at `editor.addCommand` calls, not `editor.fire` as well. oops.

list update goes well with list props. I'll use that.

File: modules/tinymce/src/plugins/lists/test/ts/browser/ListPropertiesTest.ts [hunk 5]
```diff
@@ -132,4 +134,52 @@ describe('browser.tinymce.plugins.lists.ListPropertiesTest', () => {
     await TinyUiActions.pWaitForUi(editor, '.tox-collection__item:contains("List properties"):not(.tox-collection__item--state-disabled)');
     TinyUiActions.clickOnMenu(editor, '.tox-mbtn:contains("Custom")');
   });
+
+  it('TINY-6907: List properties command is used to update the DOM', async () => {
```
Comment (JamesToohey) [inline modules/tinymce/src/plugins/lists/test/ts/browser/ListPropertiesTest.ts:138]: This test title initially confused me, if I'm understanding correctly, maybe something like:

```suggestion
  it('TINY-6907: List properties dialog uses mceListUpdate command internally to update the DOM', async () => {
```

**Review by JamesToohey — APPROVED** [review]

File: modules/tinymce/src/plugins/lists/main/ts/actions/Update.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+import { getParentList } from '../core/Selection';
+
+interface ListUpdate {
+  attrs?: Record<string, string>;
+}
+
+export const updateList = (editor: Editor, update: ListUpdate) => {
+  const parentList = getParentList(editor);
+  editor.undoManager.transact(() => {
+    if (update.attrs) {
```
Comment (ltrouton) [inline modules/tinymce/src/plugins/lists/main/ts/actions/Update.ts:19]: Should we do an `isObject` check here?

**Review by ltrouton — APPROVED** [review]

File: modules/tinymce/src/plugins/lists/main/ts/actions/Update.ts (hunk 6 above)
Comment (TheSpyder) [reply modules/tinymce/src/plugins/lists/main/ts/actions/Update.ts:19]: Indeed! I added a test for passing null `update` and it fails. Good catch!

**Comment by TheSpyder** [comment]
Looks like the remaining test failure is a safari flake, so it should be good to merge whenever you're ready 🤔

**Pull request merged**